File: packages/butler-agent/src/agent/session-workspaces/git.ts
import { existsSync, lstatSync, statSync } from "node:fs";
import { resolve } from "node:path";
import type { CommandExecutor, CommandResult } from "../../runtime/command/contracts.ts";
import { canonicalPath, samePath } from "./path.ts";
import type { WorktreeEntry } from "./contracts.ts";

export async function git(
  executor: CommandExecutor,
  cwd: string,
  args: readonly string[],
  signal?: AbortSignal,
): Promise<CommandResult> {
  return await executor.execute({
    plan: { steps: [{ executable: "git", arguments: args }] },
    cwd,
    timeoutMs: 30_000,
    signal,
  });
}

export async function resolveRepositoryAnchor(
  path: string,
  executor: CommandExecutor,
): Promise<
  | { ok: true; path: string }
  | { ok: false; code: "git_not_installed" | "git_repository_required" | "session_workspace_unavailable" }
> {
  const candidate = resolve(path);
  if (!existsSync(candidate)) return { ok: false, code: "session_workspace_unavailable" };
  try {
    if (!statSync(candidate).isDirectory()) return { ok: false, code: "git_repository_required" };
  } catch {
    return { ok: false, code: "git_repository_required" };
  }
  const result = await git(executor, candidate, ["rev-parse", "--show-toplevel"]);
  if (result.error?.code === "ENOENT") return { ok: false, code: "git_not_installed" };
  if (result.exitCode !== 0 || !result.stdout.trim()) return { ok: false, code: "git_repository_required" };
  return { ok: true, path: canonicalPath(result.stdout.trim()) };
}

export async function listWorktrees(
  executor: CommandExecutor,
  anchorPath: string,
  signal?: AbortSignal,
): Promise<
  | { ok: true; entries: WorktreeEntry[] }
  | { ok: false; code: "cancelled" | "partial_creation" | "git_not_installed" | "git_repository_required" }
> {
  const result = await git(executor, anchorPath, ["worktree", "list", "--porcelain", "-z"], signal);
  if (result.cancelled || result.timedOut) return { ok: false, code: "cancelled" };
  if (result.error?.code === "ENOENT") return { ok: false, code: "git_not_installed" };
  if (result.exitCode !== 0) return { ok: false, code: "git_repository_required" };
  return { ok: true, entries: parseWorktrees(result.stdout) };
}

export async function localBranchExists(
  executor: CommandExecutor,
  anchorPath: string,
  branch: string,
  signal?: AbortSignal,
): Promise<boolean> {
  const result = await git(executor, anchorPath, ["show-ref", "--verify", "--quiet", `refs/heads/${branch}`], signal);
  return result.exitCode === 0;
}

export async function validateLinkedWorktree(input: {
  executor: CommandExecutor;
  anchorPath: string;
  path: string;
  branch: string;
  signal?: AbortSignal;
}): Promise<
  | { ok: true; path: string; dirty: boolean }
  | { ok: false; code: "cancelled" | "partial_creation" | "linked_worktree_not_found" | "git_repository_required" | "git_not_installed" }
> {
  if (input.signal?.aborted) return { ok: false, code: "cancelled" };
  if (!existsSync(input.path)) return { ok: false, code: "linked_worktree_not_found" };
  try {
    if (lstatSync(input.path).isSymbolicLink()) return { ok: false, code: "linked_worktree_not_found" };
    const pathStat = statSync(input.path);
    if (!pathStat.isDirectory()) return { ok: false, code: "linked_worktree_not_found" };
  } catch {
    return { ok: false, code: "linked_worktree_not_found" };
  }
  const top = await git(input.executor, input.path, ["rev-parse", "--show-toplevel"], input.signal);
  if (top.cancelled || top.timedOut) return { ok: false, code: "cancelled" };
  if (top.error?.code === "ENOENT") return { ok: false, code: "git_not_installed" };
  if (top.exitCode !== 0) return { ok: false, code: "git_repository_required" };
  const worktrees = await listWorktrees(input.executor, input.anchorPath, input.signal);
  if (!worktrees.ok) return { ok: false, code: worktrees.code };
  const listed = worktrees.entries.find((entry) => samePath(entry.path, input.path) && entry.branch === input.branch);
  if (!listed) return { ok: false, code: "partial_creation" };
  const branchResult = await git(input.executor, input.path, ["symbolic-ref", "--quiet", "--short", "HEAD"], input.signal);
  if (branchResult.cancelled || branchResult.timedOut) return { ok: false, code: "cancelled" };
  if (branchResult.error?.code === "ENOENT") return { ok: false, code: "git_not_installed" };
  if (branchResult.exitCode !== 0 || branchResult.stdout.trim() !== input.branch) return { ok: false, code: "partial_creation" };
  const status = await inspectWorktreeDirty(input.executor, input.path, input.signal);
  if (!status.ok) return status;
  return { ok: true, path: canonicalPath(input.path), dirty: status.dirty };
}

export async function inspectWorktreeDirty(
  executor: CommandExecutor,
  path: string,
  signal?: AbortSignal,
): Promise<
  | { ok: true; dirty: boolean }
  | { ok: false; code: "cancelled" | "git_repository_required" | "git_not_installed" }
> {
  const status = await git(executor, path, ["status", "--porcelain=v1", "-z", "--untracked-files=all"], signal);
  if (status.cancelled || status.timedOut) return { ok: false, code: "cancelled" };
  if (status.error?.code === "ENOENT") return { ok: false, code: "git_not_installed" };
  if (status.exitCode !== 0) return { ok: false, code: "git_repository_required" };
  return { ok: true, dirty: Boolean(status.stdout) };
}

export async function inspectPartialCreation(
  executor: CommandExecutor,
  anchorPath: string,
  targetPath: string,
  branch: string,
): Promise<boolean> {
  if (!targetPath || !existsSync(targetPath)) return false;
  const listed = await listWorktrees(executor, anchorPath);
  return listed.ok && listed.entries.some((entry) => samePath(entry.path, targetPath) && entry.branch === branch);
}

function parseWorktrees(stdout: string): WorktreeEntry[] {
  const records: string[][] = [];
  let fields: string[] = [];
  for (const field of stdout.split("\0")) {
    if (!field) {
      if (fields.length > 0) records.push(fields);
      fields = [];
      continue;
    }
    fields.push(field);
  }
  if (fields.length > 0) records.push(fields);
  return records.reduce<WorktreeEntry[]>((entries, fields) => {
    const lines = fields.flatMap((field) => field.split(/\r?\n/u));
    const path = lines.find((line) => line.startsWith("worktree "))?.slice("worktree ".length);
    if (!path) return entries;
    const head = lines.find((line) => line.startsWith("HEAD "))?.slice("HEAD ".length);
    const branchRef = lines.find((line) => line.startsWith("branch "))?.slice("branch ".length);
    entries.push({
      path: canonicalPath(path),
      ...(head ? { head } : {}),
      ...(branchRef?.startsWith("refs/heads/") ? { branch: branchRef.slice("refs/heads/".length) } : {}),
    });
    return entries;
  }, []);
}

File: packages/butler-agent/rust/agent/src/workspace/session_worktree/git.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;

use tokio_util::sync::CancellationToken;

use super::path::{canonical_path, linked_directory, occupied};
use crate::workspace::WorkspaceCode;
use crate::workspace::session_recovery::path::{WorktreeEntry, parse_worktrees};
use crate::workspace::{
    CommandStep, NativeCommands, NativeWorkspaceFiles, StructuredCommandInput,
    StructuredCommandOutput, WorkspaceError, WorkspaceResult,
};

pub(super) type GitOutcome<T> = Result<T, WorkspaceCode>;

pub(super) struct Validated {
    pub path: String,
    pub dirty: bool,
}

pub(super) struct GitWorktrees<'a> {
    commands: &'a NativeCommands,
    files: &'a NativeWorkspaceFiles,
    host_environment: &'a Arc<HashMap<String, String>>,
}

impl<'a> GitWorktrees<'a> {
    pub(super) fn new(
        commands: &'a NativeCommands,
        files: &'a NativeWorkspaceFiles,
        host_environment: &'a Arc<HashMap<String, String>>,
    ) -> Self {
        Self {
            commands,
            files,
            host_environment,
        }
    }

    pub(super) async fn run(
        &self,
        cwd: &str,
        args: Vec<String>,
        abort: CancellationToken,
    ) -> WorkspaceResult<StructuredCommandOutput> {
        let input = StructuredCommandInput {
            steps: vec![CommandStep {
                executable: "git".into(),
                arguments: args,
            }],
            cwd: Some(PathBuf::from(cwd)),
            environment: HashMap::new(),
            host_environment: (**self.host_environment).clone(),
            inherit_environment: true,
            stdin: String::new(),
            timeout_ms: Some(30_000.0),
            abort,
            legacy: None,
        };
        self.commands
            .submit_structured(input)
            .map_err(WorkspaceError::from)?
            .await
            .map_err(|source| {
                WorkspaceError::new(WorkspaceCode::SessionWorktreeCommandLost, "Git result lost")
                    .with_source(source)
            })
    }

    pub(super) async fn failure(
        &self,
        cwd: &str,
        args: &[&str],
        abort: CancellationToken,
        other: WorkspaceCode,
    ) -> WorkspaceResult<Option<WorkspaceCode>> {
        let output = self
            .run(cwd, args.iter().map(|arg| (*arg).into()).collect(), abort)
            .await?;
        Ok(command_failure(&output, other))
    }

    pub(super) async fn repository_anchor(
        &self,
        path: &str,
    ) -> WorkspaceResult<GitOutcome<String>> {
        let path = path.to_owned();
        let candidate = self
            .files
            .run(move || {
                let canonical = canonical_path(&path)?;
                if !canonical.exists() {
                    return Ok::<_, std::io::Error>((None, false));
                }
                if !canonical.metadata().is_ok_and(|metadata| metadata.is_dir()) {
                    return Ok((None, true));
                }
                Ok((Some(canonical.to_string_lossy().into_owned()), false))
            })
            .await
            .map_err(WorkspaceError::from)?
            .map_err(io_error)?;
        let (candidate, not_directory) = candidate;
        let Some(candidate) = candidate else {
            return Ok(Err(if not_directory {
                WorkspaceCode::GitRepositoryRequired
            } else {
                WorkspaceCode::SessionWorkspaceUnavailable
            }));
        };
        let result = self
            .run(
                &candidate,
                vec!["rev-parse".into(), "--show-toplevel".into()],
                CancellationToken::new(),
            )
            .await?;
        if result
            .error
            .as_ref()
            .is_some_and(|error| error.code() == "ENOENT")
        {
            return Ok(Err(WorkspaceCode::GitNotInstalled));
        }
        let top = crate::public_text::trim_js_whitespace(&result.stdout);
        if result.exit_code != Some(0) || top.is_empty() {
            return Ok(Err(WorkspaceCode::GitRepositoryRequired));
        }
        let top = top.to_owned();
        let canonical = self
            .files
            .run(move || canonical_path(&top))
            .await
            .map_err(WorkspaceError::from)?
            .map_err(io_error)?;
        Ok(Ok(canonical.to_string_lossy().into_owned()))
    }

    pub(super) async fn list(
        &self,
        anchor: &str,
        abort: CancellationToken,
    ) -> WorkspaceResult<GitOutcome<Vec<WorktreeEntry>>> {
        let result = self
            .run(
                anchor,
                vec![
                    "worktree".into(),
                    "list".into(),
                    "--porcelain".into(),
                    "-z".into(),
                ],
                abort,
            )
            .await?;
        if let Some(code) = command_failure(&result, WorkspaceCode::GitRepositoryRequired) {
            return Ok(Err(code));
        }
        let entries = self
            .files
            .run(move || parse_worktrees(&result.stdout))
            .await
            .map_err(WorkspaceError::from)?
            .map_err(io_error)?;
        Ok(Ok(entries))
    }

    pub(super) async fn same_path(&self, left: &str, right: &str) -> WorkspaceResult<bool> {
        let left = left.to_owned();
        let right = right.to_owned();
        self.files
            .run(move || Ok::<_, std::io::Error>(canonical_path(&left)? == canonical_path(&right)?))
            .await
            .map_err(WorkspaceError::from)?
            .map_err(io_error)
    }

    pub(super) async fn occupied(&self, path: &str) -> WorkspaceResult<bool> {
        let path = path.to_owned();
        self.files
            .run(move || occupied(&path))
            .await
            .map_err(WorkspaceError::from)
    }

    pub(super) async fn local_branch_exists(
        &self,
        anchor: &str,
        branch: &str,
        abort: CancellationToken,
    ) -> WorkspaceResult<bool> {
        let result = self
            .run(
                anchor,
                vec![
                    "show-ref".into(),
                    "--verify".into(),
                    "--quiet".into(),
                    format!("refs/heads/{branch}"),
                ],
                abort,
            )
            .await?;
        Ok(result.exit_code == Some(0))
    }

    pub(super) async fn validate(
        &self,
        anchor: &str,
        target: &str,
        branch: &str,
        abort: CancellationToken,
    ) -> WorkspaceResult<GitOutcome<Validated>> {
        if abort.is_cancelled() {
            return Ok(Err(WorkspaceCode::Cancelled));
        }
        let path = target.to_owned();
        if !self
            .files
            .run(move || linked_directory(&path))
            .await
            .map_err(WorkspaceError::from)?
        {
            return Ok(Err(WorkspaceCode::LinkedWorktreeNotFound));
        }
        let top = self
            .run(
                target,
                vec!["rev-parse".into(), "--show-toplevel".into()],
                abort.clone(),
            )
            .await?;
        if let Some(code) = command_failure(&top, WorkspaceCode::GitRepositoryRequired) {
            return Ok(Err(code));
        }
        let listed = match self.list(anchor, abort.clone()).await? {
            Ok(value) => value,
            Err(code) => return Ok(Err(code)),
        };
        let mut found = false;
        for entry in listed {
            if entry.branch.as_deref() == Some(branch)
                && self
                    .same_path(&entry.path.to_string_lossy(), target)
                    .await?
            {
                found = true;
                break;
            }
        }
        if !found {
            return Ok(Err(WorkspaceCode::PartialCreation));
        }
        let symbolic = self
            .run(
                target,
                vec![
                    "symbolic-ref".into(),
                    "--quiet".into(),
                    "--short".into(),
                    "HEAD".into(),
                ],
                abort.clone(),
            )
            .await?;
        if let Some(code) = command_failure(&symbolic, WorkspaceCode::PartialCreation) {
            return Ok(Err(code));
        }
        if crate::public_text::trim_js_whitespace(&symbolic.stdout) != branch {
            return Ok(Err(WorkspaceCode::PartialCreation));
        }
        let dirty = match self.dirty(target, abort).await? {
            Ok(value) => value,
            Err(code) => return Ok(Err(code)),
        };
        let path = target.to_owned();
        let canonical = self
            .files
            .run(move || canonical_path(&path))
            .await
            .map_err(WorkspaceError::from)?
            .map_err(io_error)?;
        Ok(Ok(Validated {
            path: canonical.to_string_lossy().into_owned(),
            dirty,
        }))
    }

    pub(super) async fn dirty(
        &self,
        path: &str,
        abort: CancellationToken,
    ) -> WorkspaceResult<GitOutcome<bool>> {
        let status = self
            .run(
                path,
                vec![
                    "status".into(),
                    "--porcelain=v1".into(),
                    "-z".into(),
                    "--untracked-files=all".into(),
                ],
                abort,
            )
            .await?;
        if let Some(code) = command_failure(&status, WorkspaceCode::GitRepositoryRequired) {
            return Ok(Err(code));
        }
        Ok(Ok(!status.stdout.is_empty()))
    }

    pub(super) async fn partial_creation(
        &self,
        anchor: &str,
        target: &str,
        branch: &str,
    ) -> WorkspaceResult<bool> {
        let target_path = target.to_owned();
        if !self
            .files
            .run(move || std::path::Path::new(&target_path).exists())
            .await
            .map_err(WorkspaceError::from)?
        {
            return Ok(false);
        }
        let Ok(entries) = self.list(anchor, CancellationToken::new()).await? else {
            return Ok(false);
        };
        for entry in entries {
            if entry.branch.as_deref() == Some(branch)
                && self
                    .same_path(&entry.path.to_string_lossy(), target)
                    .await?
            {
                return Ok(true);
            }
        }
        Ok(false)
    }
}

pub(super) fn command_failure(
    result: &StructuredCommandOutput,
    other: WorkspaceCode,
) -> Option<WorkspaceCode> {
    if result.cancelled || result.timed_out {
        Some(WorkspaceCode::Cancelled)
    } else if result
        .error
        .as_ref()
        .is_some_and(|error| error.code() == "ENOENT")
    {
        Some(WorkspaceCode::GitNotInstalled)
    } else if result.exit_code != Some(0) {
        Some(other)
    } else {
        None
    }
}

fn io_error(error: std::io::Error) -> WorkspaceError {
    WorkspaceError::new(WorkspaceCode::SessionWorktreeIo, error.to_string()).with_source(error)
}
